File: dsa-problems/ProbabilisticDS/solutions/cpp/PDS-010-count-sketch-heavy-hitters.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int countSketchEstimate(const vector<int>& count, const vector<int>& sign) {
        int d = count.size();
        vector<int> estimates(d);
        for (int i = 0; i < d; i++) {
            estimates[i] = count[i] * sign[i];
        }
        sort(estimates.begin(), estimates.end());
        return estimates[d / 2];
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int d;
    if (cin >> d) {
        vector<int> count(d), sign(d);
        for (int i = 0; i < d; i++) {
            cin >> count[i] >> sign[i];
        }
    
        Solution solution;
        cout << solution.countSketchEstimate(count, sign) << "\n";
    }
    return 0;
}

File: dsa-problems/ProbabilisticDS/solutions/java/PDS-010-count-sketch-heavy-hitters.java
import java.util.*;

class Solution {
    public int countSketchEstimate(int[] count, int[] sign) {
        int d = count.length;
        int[] estimates = new int[d];
        for (int i = 0; i < d; i++) {
            estimates[i] = count[i] * sign[i];
        }
        Arrays.sort(estimates);
        return estimates[d / 2];
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (sc.hasNextInt()) {
            int d = sc.nextInt();
            int[] count = new int[d];
            int[] sign = new int[d];
            for (int i = 0; i < d; i++) {
                count[i] = sc.nextInt();
                sign[i] = sc.nextInt();
            }
    
            Solution solution = new Solution();
            System.out.println(solution.countSketchEstimate(count, sign));
        }
        sc.close();
    }
}
